src/sync/app.cpp

```diff
@@ -0,0 +1,188 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or utilied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#include "app.hpp"
+#include <sstream>
+#include "sync_manager.hpp"
+#include "generic_network_transport.hpp"
+#include "app_credentials.hpp"
+#include "../external/json/json.hpp"
+
+// wrap an optional json key into the Optional type
+#define WRAP_JSON_OPT(JSON, KEY, RET_TYPE) \
+JSON.find(KEY) != JSON.end() ? Optional<RET_TYPE>(JSON[KEY].get<RET_TYPE>()) : realm::util::none
+
+namespace realm {
+namespace app {
+
+std::shared_ptr<App> App::app(const std::string app_id,
+                              const realm::util::Optional<App::Config> config)
+{
+    return std::make_shared<App>(app_id, config);
+}
+
+const std::string App::default_base_url = "https://stitch.mongodb.com";
+const std::string App::base_path = "/api/client/v2.0";
+const std::string App::app_path = "/app";
+const std::string App::auth_path = "/auth";
+const uint64_t App::default_timeout_ms = 60000;
+
+static std::unique_ptr<error::AppError> handle_error(const Response response) {
+    if (response.body.empty()) {
+        return {};
+    }
+
+    if (response.headers.find("Content-Type") != response.headers.end()) {
+        if (response.headers.at("Content-Type") == "application/json") {
+            auto body = nlohmann::json::parse(response.body);
+            return std::make_unique<error::AppError>(app::error::ServiceError(body["errorCode"].get<std::string>(),
+                                  body["error"].get<std::string>()));
+        }
+    }
+
+    if (response.custom_status_code != 0) {
+        return std::make_unique<error::AppError>(response.body,
+                                                 response.custom_status_code,
+                                                 error::AppError::Type::Custom);
+    }
+
+    return {};
+}
+
+inline bool response_code_is_fatal(Response response) {
+    // FIXME: our tests currently only generate codes 0 and 200,
+    // but we need more robust error handling here; eg. should a 300
+    // redirect really be considered fatal or should we automatically redirect?
+    return response.http_status_code >= 300 || response.http_status_code < 200 || response.custom_status_code != 0;
+}
+
+void App::login_with_credentials(const std::shared_ptr<AppCredentials> credentials,
+                                 std::function<void(std::shared_ptr<SyncUser>, std::unique_ptr<error::AppError>)> completion_block) {
```

Comment:
This completion function could pass `util::Optional<AppError>` and simplify all the code around building errors. Despite having subclasses, `AppError` isn't actually a polymorphic type.